stm32wb55/rfcore: Protect against file errors in rfcore firmware update scripts.

andrewleech opened the pull request:
We've run into a few failures in the wb55 rfcore update procedures lately - mostly caused with the stm gui application, but sometimes from un-recognised failures during bin file transfer to the wb55 prior to running the `rfcore_firmware.py` script - causing corrupted firmware to be flashed to the unit.

Other failures were caused by incorrect load addresses being used, again both from user error copying the address from the html release notes to the gui tool, but also from similarly not updating the address correctly in `rfcore_firmware.py`

To guard against these errors and make it easier to prepare different versions I've added a few features to the rfcore firmware tools:
* When creating the bin file, automatically parse the release note in the folder to get the correct address.
* Add a footer to the bin file containing the name, version, crc, address etc.
* Before flashing rfcore, check if the same version is already installed.
* Verify the crc and obfuscation key before flashing bin.
* log the name and version of file being flashed.

jimmo commented on ports/stm32/boards/NUCLEO_WB55/rfcore_firmware.py:
```diff
@@ -339,8 +414,17 @@ def _write_failure_state(reason):
+
+def _write_addr(addr):
+    machine.mem32[REG_RTC_ADDR] = addr
+    return addr
```
The return value is unused.

andrewleech replied:
True, I was copying the pattern of _write_failure_state above but yeah it's return is used at one point. _write_state above that doesn't return anything. So it wasn't really a pattern at all...

jimmo commented on ports/stm32/boards/NUCLEO_WB55/rfcore_firmware.py:
```diff
@@ -339,8 +414,17 @@ def _write_failure_state(reason):
     return reason
 
 
+def _read_addr():
```
Would be good to come up with a better name for this... maybe `{read,write}_target_addr` ?

jimmo commented on ports/stm32/boards/NUCLEO_WB55/rfcore_firmware.py:
```diff
@@ -558,7 +648,19 @@ def resume():
-    log("Starting firmware update")
-    _write_state(_STATE_WAITING_FOR_FUS)
-    machine.reset()
+def check_for_updates(force=False):
```
Could you be doing a FUS-only update? In which case checking the WS might prevent that from running (unless `force` is set)?

andrewleech replied:
Good point. It "feels" unlikely to me that there would be a board in a state where it has up to date WUS but older FUS, but that's certainly possible. I should check for FUS version as well perhaps

jimmo commented on ports/stm32/boards/NUCLEO_WB55/rfcore_makefirmware.py:
```diff
@@ -41,27 +44,149 @@
+        check_file_details(dest_file)
+
+
+def validate_crc(f):
```
I'm not sure it's that important but would be good to avoid duplication... We could put these in `rfcore_firmware_utils.py` and make sure both get frozen?

andrewleech replied:
rfcore_firmware.py could be refactored slightly to allow it to be imported into rfcore_makefirmware, then it could be single source of truth for a few things (this function, the KEY, etc)

jimmo commented on ports/stm32/boards/NUCLEO_WB55/rfcore_makefirmware.py:
```diff
@@ -41,27 +44,149 @@
+        print(
+            "Cannot determine binary load address, please confirm manually in Release Notes."
+        )
+        release_notes = None
```
Should this terminate? It looks like `release_notes` being None will just crash when the return value of `get_details` is unpacked.

jimmo commented on ports/stm32/boards/NUCLEO_WB55/rfcore_firmware.py:
```diff
@@ -308,17 +380,20 @@ def _fus_fwdelete():
-def _fus_run_fwupgrade(addr):
+def _fus_run_fwupgrade():
     # Note: Address is ignored by the FUS (see comments above).
+    addr = _read_addr()
```
Should this verify `addr` ? (Specifically that it's non-zero, mostly just as a sanity check)

jimmo commented on ports/stm32/boards/NUCLEO_WB55/rfcore_makefirmware.py:
```diff
@@ -29,9 +29,12 @@
 # an update. See more information (and the corresponding de-obfuscation) in
 # rfcore_firmware.py as well as instructions on how to use.
 
+from imp import release_lock
```
Unused?

andrewleech replied:
Oh, I think vscode might have added that automatically, shouldn't be there.

jimmo reviewed (commented):
Thanks @andrewleech this all looks really good.

Just a minor note that there seems to be additional formatting changes that are different to what the CI version of `black` recommends. Perhaps you have a later version. And (sorry) super nitpick for consistency that in the rest of this file all the comments end in full stops.

jimmo commented:
I wondered if maybe the footer should have some sort of magic identifier, but I guess in a sense the KEY check serves that purpose. However, mostly just out of habit, maybe we want to put some sort of version identifier the footer (i.e. footer format version, not firmware version)? I guess we could just change the key.

pi-anl committed:
stm32wb55/rfcore: Error handling for rfcore fw update scripts.

andrewleech commented:
I've addressed the earlier review feedback, however final re-testing is still pending (on my end).

andrewleech commented:
Rebased and tested with [rfcore v1.14.0](https://github.com/STMicroelectronics/STM32CubeWB/releases/tag/v1.14.0) on WB55 dongle.

I also added an extra function to automate adding the required call to boot.py and expanded the usage instructions a little in 
rfcore_firmware.py

dpgeorge commented:
Rebased and merged in 6804b6f54fbed1f0fdb86668ebb46274e4876fa6

Thanks!
